Merge interfaces

MikeVezina opened the pull request:
#### Problem
- Needed to make changes to original code base since the way the sensor communicates data is not what was originally anticipated.

#### Solution
- Changed code base to include all necessary changes. 
- Tests need to be updated.
- Some files have also been updated to conform to PEP8 standards with respect to whitespace and formatting.
- There is also a sensor stub that allows us to send fake sensor data as a way to debug our end-to-end system without a physical sensor.

 - [x] Tested locally?

erickadbay commented on services/api_service.py:
```diff
@@ -1,44 +1,45 @@
+        return cls.__notify('post', '/desks.json', {'desk': {'name': name}})
+
+    @classmethod
+    def update_desk_occupied_status(cls, desk_id, occupied_status):
```
You need to tweak `api_service_test.py` to accommodate for this change

MikeVezina replied:
Thank you!

erickadbay commented on desk.py:
```diff
@@ -68,16 +68,18 @@ def remove_observer(self, obj):
         self.__desk_observers.discard(obj)
 
-    def input_received(self, pin, value):
+    def input_received(self, value):
```
Fyi some tests are failing because of this change

erickadbay commented on config/config.py:
```diff
@@ -1,187 +1,240 @@
+    def __init__(self, desk):
+        self.__desk_id = desk[self.KEY_ID]
+        self.__desk_name = desk[self.KEY_NAME]
+        self.__desk_sensor = _DeskSensorConfigDAO(desk[self.KEY_SENSOR])
```
Why not send the whole desk to `_DeskSensorConfigDAO`? That way `DeskHandlerService` doesnt have to send redundant parameters

MikeVezina commented:
- I changed the circle.yml config file so that we override the working directory when running the tests. This allows us to pick up the proper configuration file under the tests/ directory, without messing up any of the imports. (Also, we no longer rely on environment variables to determine which directory we load the configuration from)

- Added a bash script that allows us to run all tests in a similar way to how Circle CI would run them

- Ive made all the changes to the tests, they all pass now and properly reflect what they should be testing.

- Made some changes to ConfigDAO so that we can re-load the config from file
- We can also modify the config and save it.
- There is now a function in ConfigDAO that allows us to add a desk without having to pass in a Dictionary object
- The config also saves the i2c_address as a string (so it is a readable hex value rather than a base 10 integer). DeskSensorConfigDAO properly handles the conversion from hex string to int.

DominikSchmidtlein commented on desk_sensor_table.py:
```diff
@@ -0,0 +1,67 @@
+        """
+
+        # Get all desks config
+        all_desks = CONFIG.desks
```
If `all_desks` isn't used more than once, I wouldn't create it.

DominikSchmidtlein commented on desk_sensor_table.py:
```diff
@@ -0,0 +1,67 @@
+    # The purpose of desk id in the tuple is used for easy lookup of desks
+    # The sensor object is the object responsible for reading sensor data and reporting occupancy status
+    # The desk object is used for performing additional operations on the desk if necessary
+    __desk_mapping = []
```
Could this be a dictionary with `id` as key? It would make `get_mapping_from_desk_id` a lot easier.

DominikSchmidtlein commented on desk_sensor_table.py:
```diff
@@ -0,0 +1,67 @@
+            self.__desk_mapping.append((d.id, sensor_obj, Desk(d.name, d.id)))
+
+            # Print additional information in debug mode
+            if CONFIG.debug_config.enabled:
```
This sort of behavior (conditional logging) could be handled by the logger if we choose to use a logging library.

MikeVezina replied:
Yep. Ill have to start adding that into our code base

DominikSchmidtlein commented on desk_sensor_table.py:
```diff
@@ -0,0 +1,67 @@
+
+            # Print additional information in debug mode
+            if CONFIG.debug_config.enabled:
+                print("Added Desk to DeskPinTable: [ID]:", d.id, "[Name]:", d.name, "[Sensor Type]:", d.sensor.sensor_type)
```
`String.format()` is a more efficient way to format strings like this.

DominikSchmidtlein replied:
`f"{var}"` is even cleaner

DominikSchmidtlein commented on desk_sensor_table.py:
```diff
@@ -0,0 +1,67 @@
+    def cleanup(self):
+        # Close all i2c connections
+        for (desk_id, sensor_obj, desk_obj) in self.__desk_mapping:
+            print("Closing i2c connection for desk: " + str(desk_id))
```
`String.format()`

DominikSchmidtlein replied:
`f"{var}"` is even cleaner

DominikSchmidtlein commented on sensors/omron_d6t/omron_d6t.py:
```diff
@@ -38,15 +44,15 @@ def __read(self):
 
         # This loop will run until a valid value is received from the sensor (or max retries is reached)
-        for i in range(0, self.MAX_RETRIES):
+        for i in range(0, self._MAX_RETRIES):
```
`range` defaults to starting at 0. `range(self._MAX_RETRIES)` is sufficient

DominikSchmidtlein commented on sensors/omron_d6t/omron_d6t.py:
```diff
@@ -38,15 +44,15 @@ def __read(self):
 
             if bytes_read == self.__BUFFER_LENGTH:
                 # Read PEC byte for checksum
                 crc_expected = sensor_raw_data[34]
```
Magic number

MikeVezina replied:
I made changes to use -1, but is -1 still considered a magic number?

DominikSchmidtlein commented on sensors/omron_d6t/omron_d6t.py:
```diff
@@ -83,11 +90,14 @@ def __parse_temperature_data(self, raw_sensor_buffer):
         # See OMRON datasheet for PEC details
-        pec_buffer = [self.__PEC_PREFIX]
+        for i in self.__PEC_PREFIX:
+            pec_buffer.append(i)
```
1 liner: `pec_buffer.extend(self.__PEC_PREFIX)`

DominikSchmidtlein commented on sensors/omron_d6t/omron_d6t.py:
```diff
@@ -83,11 +90,14 @@ def __parse_temperature_data(self, raw_sensor_buffer):
+            pec_buffer.append(i)
 
         for itm in buf:
             pec_buffer.append(itm)
```
`List#extend`

MikeVezina replied:
This is awesome. Thank you!

DominikSchmidtlein commented on sensors/sensor.py:
```diff
@@ -45,15 +43,39 @@ def occupied_status(self):
+                else:
+                    exit(-1)
 
+            print("Successfully Connected! (Version: " + str(Sensor.__pi_gpio_instance.get_pigpio_version()) + ")\n")
```
`String#format`

DominikSchmidtlein replied:
I think it would be good practice to have a `getInstance` method for `__pi_gpio_instance`

DominikSchmidtlein replied:
`f"{var}"` is even cleaner

MikeVezina replied:
I always forget that python has this, it's such a nice convenience!

erickadbay commented on sensors/omron_d6t/omron_d6t.py:
```diff
@@ -83,13 +90,13 @@ def __parse_temperature_data(self, raw_sensor_buffer):
-            pec_buffer.append(itm)
+        # See OMRON datasheet for PEC details
+        pec_buffer.extend(self.__PEC_PREFIX)
+        pec_buffer.extend(buf)
```
```python
pec_buffer.extend([self.__PEC_PREFIX, buf])
```

MikeVezina replied:
This appends each array as an item in pec_buffer. So you would get something like this

pec_buffer = [ self.__PEC_PREFIX, buf ]
pec_buffer = [[...], [...]] 

Instead of actually appending the individual items of PEC_PREFIX and buf into pec_buffer

erickadbay replied:
Oh. Any reason on why you need actual arrays as items?

MikeVezina replied:
Sorry! What I meant was the comment you supplied is what leads to having the arrays as items (which is not what we want!)

erickadbay replied:
Are you sure? What you said is true for `append` but not for extend from what I read 😕 
![image](https://user-images.githubusercontent.com/9994372/36236057-0de8a88a-11c2-11e8-806d-1aafad81f354.png)

MikeVezina replied:
Maybe this clears it up a bit: 

![screen shot 2018-02-14 at 8 10 12 pm](https://user-images.githubusercontent.com/10552190/36236253-1f367ada-11c3-11e8-80c1-ed63f72eee26.png)

erickadbay replied:
Oh it's because `_PEC_PREFIX` and `buf` are lists as well.

Works for me then 👍

erickadbay commented on config/config.py:
```diff
@@ -1,187 +1,330 @@
-	@property
-	def enabled(self):
-		return self.__enabled
+    @sensor_type.setter
```
Do we need all these setters?

MikeVezina replied:
Not at the moment, but I was kind of future-proofing in a way, since I was already fixing the config stuff. This way we can modify the config via CLI, if needed, and it could be easily saved to file. If it ends up truly being useless it wouldn't be hard to remove.

erickadbay commented on services/desk_handler_service.py:
```diff
@@ -23,8 +23,12 @@ def create(cls, name, i2c_address):
         json_response = response.json()
+
+        if type(i2c_address) is str:
+            i2c_address = int(i2c_address, 16)
```
👌
